src/main/java/com/example/Library/services/BookService.java

```diff
@@ -69,4 +73,26 @@ public Optional<Book> returnBook(long id){
                     return bookRepository.save(book);
                 });
     }
+
+    public Book filter(String idOrTitle) {
+        char[] chars = idOrTitle.toCharArray();
+        for (char symbol : chars) {
+            int x = symbol;
+            if ((x >= 97 && x <= 122) || (x >= 65 && x <= 90) || (x >= 1040 && x <= 1071) || (x >= 1073 && x <= 1103)) {
+                return bookRepository.getBookByTitle(idOrTitle)
+                        .orElseThrow(() -> new BookNotFoundByTitleException(idOrTitle));
+            }
+        }
+        long id = Long.parseLong(idOrTitle);
+        return bookRepository.findById(id)
+                .orElseThrow(() -> new BookNotFoundByIdException(id));
+    }
+
+    public Iterable<Book> getAllBooksWithFilterByTitle(String title) {
+        if (title != null) {
+            return new ArrayList<>(Set.of(bookRepository.getBookByTitle(title)
```

Comment:
there may be several books with the same title (as it is not a unique idendifier), for this reason, it would be better to update bookRepository.getBookByTitle to return a collections of books instead of Optional<Book>, and this transformation to Set will not be needed